File: amethyst/tools/isobuilder/iso/isofile.py
import os
from .binaryfile import BinaryFile

class IsoFile:
    """A file in an ISO."""

    def __init__(self, path:str, isDir:bool=False, offset:int=0,
    size:int=0, file:BinaryFile=None, fileOffs:int=0, parent=None,
    isSystem=False):
        """Create file in ISO.

        path:     file path within ISO.
        isDir:    whether this represents a directory (True) or file (False).
        offset:   file data offset within ISO.
        size:     file data size in bytes.
        file:     ISO or data file to read from.
        fileOffs: offset to read data from within file.
        parent:   parent IsoFile.
        isSystem: whether this is a system file (extracted to /sys
            instead of /files).
        """
        self.path     = path
        self.isDir    = isDir
        self.offset   = offset
        self.size     = size
        self.file     = file
        self.fileOffs = fileOffs
        self.parent   = parent
        self.isSystem = False # is system file?


    def __str__(self):
        return '<ISO %s "%s" @ 0x%x>' % (
            'Directory' if self.isDir else 'File',
            self.path, id(self))


    def isDescendantOf(self, file):
        """Check if this file is a child or descendant of the given file."""
        return (self.parent == file) or (self.parent is not None and self.parent.isDescendantOf(file))


    def replaceWith(self, file:(str,BinaryFile), offset:int=0, size:int=None):
        """Replace this file with one from the local disk.

        file:   file to replace with.
        offset: offset within file to read from.
        size:   data size. if None, will use file size - offset.
        """
        #if type(file) is str: file = BinaryFile(file, 'rb')
        self.file     = file
        self.fileOffs = offset
        if size is None:
            if type(file) is str:
                size = os.stat(file).st_size - offset
            else:
                file.seek(offset)
                size = file.size - offset
        self.size = size


    def writeToFile(self, file:(str,BinaryFile), chunkSize:int=4096):
        """Write this file's content to disk."""
        if self.isDir: raise IsADirectoryError(self)
        if not self.file:
            raise FileNotFoundError("no source file to read from")

        inFile = self.file
        if type(inFile) is str: inFile = BinaryFile(inFile, 'rb')

        if type(file) is str: file = BinaryFile(file, 'wb')
        #print("file %s at 0x%08X size 0x%08X in %s" % (
        #    self.path, self.offset, self.size, file))
        inFile.seek(self.fileOffs)
        for i in range(0, self.size, chunkSize):
            size = min(chunkSize, self.size - i)
            data = inFile.read(size)
            if len(data) != size:
                raise RuntimeError("Expected 0x%X bytes but got 0x%X at 0x%X in %s" % (
                    size, len(data), inFile.tell(), self.path))
            file.write(data)

File: browser/r/js/types/iso/isofile.js
import Struct from '../../lib/Struct.js';

export default class IsoFile {
    //A file in an ISO.

    constructor(path, isDir=false, offset=0, size=0, buffer=null,
    bufferOffs=0, parent=null, isSystem=false) {
        /** @description Create file in ISO.
         *  @param path       file path within ISO.
         *  @param isDir      whether this represents a directory (true) or file (false).
         *  @param offset     file data offset within ISO.
         *  @param size       file data size in bytes.
         *  @param buffer     buffer to read from.
         *  @param bufferOffs offset to read data from within buffer.
         *  @param parent     parent IsoFile.
         *  @param isSystem   whether this is a system file (extracted to /sys
         *      instead of /files).
         */
        this.path       = path;
        this.isDir      = isDir;
        this.offset     = offset;
        this.size       = size;
        this.buffer     = buffer;
        this.bufferOffs = bufferOffs;
        this.parent     = parent;
        this.isSystem   = isSystem;
        this.name       = path.split('/');
        this.name       = this.name[this.name.length - 1];
        this._data      = {}; //cache
    }

    isDescendantOf(file) {
        //Check if this file is a child or descendant of the given file.
        return ((self.parent == file) ||
            (self.parent != null && self.parent.isDescendantOf(file)));
    }

    getFormat(offset=0) {
        let view = new DataView(this.buffer, this.bufferOffs+offset, 4);
        const magic = view.getUint32(0, false);
        switch(magic) {
            case 0x5A4C4200: return "ZLB"; //'ZLB\0'
            default: return "raw";
        }
    }

    getData(offset=0, size=0) {
        if(size <= 0) size += this.size;

        let key = `R${offset},${size}`;
        if(!this._data[key]) {
            this._data[key] = new DataView(this.buffer,
                this.bufferOffs+offset, size);
        }
        return this._data[key];
    }
};
